File: src/main/java/net/dontdrinkandroot/gitki/wicket/component/modal/MoveFileModal.java
package net.dontdrinkandroot.gitki.wicket.component.modal;

import net.dontdrinkandroot.gitki.model.DirectoryPath;
import net.dontdrinkandroot.gitki.model.FilePath;
import net.dontdrinkandroot.gitki.model.Role;
import net.dontdrinkandroot.gitki.service.git.GitService;
import net.dontdrinkandroot.gitki.wicket.GitkiWebSession;
import net.dontdrinkandroot.gitki.wicket.choicerenderer.AbstractPathAbsoluteStringChoiceRenderer;
import net.dontdrinkandroot.gitki.wicket.event.FileMovedEvent;
import net.dontdrinkandroot.gitki.wicket.security.Instantiate;
import net.dontdrinkandroot.wicket.behavior.OnClickScriptBehavior;
import net.dontdrinkandroot.wicket.bootstrap.behavior.ButtonBehavior;
import net.dontdrinkandroot.wicket.bootstrap.component.button.SubmitLabelButton;
import net.dontdrinkandroot.wicket.bootstrap.component.form.formgroup.FormGroupInputText;
import net.dontdrinkandroot.wicket.bootstrap.component.form.formgroup.FormGroupSelect;
import net.dontdrinkandroot.wicket.bootstrap.component.modal.AjaxFormModal;
import org.apache.wicket.WicketRuntimeException;
import org.apache.wicket.ajax.AjaxRequestTarget;
import org.apache.wicket.event.Broadcast;
import org.apache.wicket.markup.html.basic.Label;
import org.apache.wicket.markup.repeater.RepeatingView;
import org.apache.wicket.model.IModel;
import org.apache.wicket.model.Model;
import org.apache.wicket.model.StringResourceModel;
import org.apache.wicket.spring.injection.annot.SpringBean;
import org.eclipse.jgit.api.errors.GitAPIException;

import java.io.IOException;
import java.util.List;

/**
 * @author Philip Washington Sorst <philip@sorst.net>
 */
@Instantiate(Role.COMMITTER)
public class MoveFileModal extends AjaxFormModal<FilePath>
{
    @SpringBean
    private GitService gitService;

    private IModel<String> commitMessageModel;

    private IModel<String> targetNameModel;

    private IModel<DirectoryPath> targetDirectoryModel;

    public MoveFileModal(String id, IModel<FilePath> model)
    {
        super(id, model);
        this.commitMessageModel = Model.of("Moving " + this.getModelObject().toAbsoluteString());
        this.targetNameModel = Model.of(this.getModelObject().getName());
        this.targetDirectoryModel = Model.of(this.getModelObject().getParent());
    }

    @Override
    protected IModel<String> createHeadingModel()
    {
        return new StringResourceModel("gitki.move");
    }

    @Override
    protected void populateFormGroups(RepeatingView formGroupView)
    {
        super.populateFormGroups(formGroupView);

        List<DirectoryPath> availableDirectories = null;
        try {
            availableDirectories = this.gitService.listAllDirectories();
        } catch (IOException e) {
            throw new WicketRuntimeException(e);
        }

        FormGroupInputText formGroupTargetName =
                new FormGroupInputText(
                        formGroupView.newChildId(),
                        new StringResourceModel("gitki.name"),
                        this.targetNameModel
                );
        formGroupTargetName.setRequired(true);
        formGroupView.add(formGroupTargetName);

        FormGroupSelect<DirectoryPath> formGroupTargetPath =
                new FormGroupSelect<>(
                        formGroupView.newChildId(),
                        new StringResourceModel("gitki.targetpath"),
                        this.targetDirectoryModel,
                        availableDirectories,
                        new AbstractPathAbsoluteStringChoiceRenderer<>()
                );
        formGroupTargetPath.setRequired(true);
        formGroupView.add(formGroupTargetPath);

        FormGroupInputText formGroupCommitMessage =
                new FormGroupInputText(
                        formGroupView.newChildId(),
                        new StringResourceModel("gitki.commitmessage"),
                        this.commitMessageModel
                );
        formGroupCommitMessage.addDefaultAjaxInputValidation();
        formGroupCommitMessage.setRequired(true);
        formGroupView.add(formGroupCommitMessage);
    }

    @Override
    protected void populateFormActions(RepeatingView formActionView)
    {
        super.populateFormActions(formActionView);

        formActionView.add(formActionView.add(new SubmitLabelButton(
                formActionView.newChildId(),
                new StringResourceModel("gitki.move")
        )));

        Label cancelButton = new Label(formActionView.newChildId(), new StringResourceModel("gitki.cancel"));
        cancelButton.add(new ButtonBehavior());
        cancelButton.add(new OnClickScriptBehavior(this.getHideScript()));
        formActionView.add(cancelButton);
    }

    protected FilePath getTargetPath()
    {
        return this.targetDirectoryModel.getObject().appendFileName(this.targetNameModel.getObject());
    }

    protected void onFileMoved(AjaxRequestTarget target, FilePath sourcePath, FilePath targetPath)
    {
        this.send(this.getPage(), Broadcast.BREADTH, new FileMovedEvent(sourcePath, target, targetPath));
    }

    @Override
    protected void onSubmit(AjaxRequestTarget target)
    {
        super.onSubmit(target);

        try {
            MoveFileModal.this.gitService.moveAndCommit(
                    MoveFileModal.this.getModelObject(),
                    MoveFileModal.this.getTargetPath(),
                    GitkiWebSession.get().getUser(),
                    MoveFileModal.this.commitMessageModel.getObject()
            );
        } catch (IOException | GitAPIException e) {
            throw new WicketRuntimeException(e);
        }
    }

    @Override
    protected void onAfterSubmit(AjaxRequestTarget target)
    {
        super.onAfterSubmit(target);

        MoveFileModal.this.onFileMoved(target, MoveFileModal.this.getModelObject(), MoveFileModal.this.getTargetPath());
    }
}

File: src/main/kotlin/net/dontdrinkandroot/gitki/wicket/component/modal/MoveFileModal.kt
package net.dontdrinkandroot.gitki.wicket.component.modal

import net.dontdrinkandroot.gitki.model.DirectoryPath
import net.dontdrinkandroot.gitki.model.FilePath
import net.dontdrinkandroot.gitki.model.Role
import net.dontdrinkandroot.gitki.service.git.GitService
import net.dontdrinkandroot.gitki.wicket.choicerenderer.AbstractPathAbsoluteStringChoiceRenderer
import net.dontdrinkandroot.gitki.wicket.event.FileMovedEvent
import net.dontdrinkandroot.gitki.wicket.getGitkiSession
import net.dontdrinkandroot.gitki.wicket.security.Instantiate
import net.dontdrinkandroot.wicket.behavior.OnClickScriptBehavior
import net.dontdrinkandroot.wicket.bootstrap.behavior.ButtonBehavior
import net.dontdrinkandroot.wicket.bootstrap.component.button.SubmitLabelButton
import net.dontdrinkandroot.wicket.bootstrap.component.form.formgroup.FormGroupInputText
import net.dontdrinkandroot.wicket.bootstrap.component.form.formgroup.FormGroupSelect
import net.dontdrinkandroot.wicket.bootstrap.component.modal.AjaxFormModal
import org.apache.wicket.WicketRuntimeException
import org.apache.wicket.ajax.AjaxRequestTarget
import org.apache.wicket.event.Broadcast
import org.apache.wicket.markup.html.basic.Label
import org.apache.wicket.markup.repeater.RepeatingView
import org.apache.wicket.model.IModel
import org.apache.wicket.model.Model
import org.apache.wicket.model.StringResourceModel
import org.apache.wicket.spring.injection.annot.SpringBean
import org.eclipse.jgit.api.errors.GitAPIException
import java.io.IOException

@Instantiate(Role.COMMITTER)
class MoveFileModal(id: String, model: IModel<FilePath>) : AjaxFormModal<FilePath>(id, model) {

    @SpringBean
    private val gitService: GitService? = null
    private val commitMessageModel: IModel<String>
    private val targetNameModel: IModel<String?>
    private val targetDirectoryModel: IModel<DirectoryPath?>
    override fun createHeadingModel(): IModel<String> {
        return StringResourceModel("gitki.move")
    }

    override fun populateFormGroups(formGroupView: RepeatingView) {
        super.populateFormGroups(formGroupView)
        var availableDirectories: List<DirectoryPath?>? = null
        availableDirectories = try {
            gitService!!.listAllDirectories()
        } catch (e: IOException) {
            throw WicketRuntimeException(e)
        }
        val formGroupTargetName = FormGroupInputText(
            formGroupView.newChildId(),
            StringResourceModel("gitki.name"),
            targetNameModel
        )
        formGroupTargetName.setRequired(true)
        formGroupView.add(formGroupTargetName)
        val formGroupTargetPath = FormGroupSelect(
            formGroupView.newChildId(),
            StringResourceModel("gitki.targetpath"),
            targetDirectoryModel,
            availableDirectories,
            AbstractPathAbsoluteStringChoiceRenderer<DirectoryPath?>()
        )
        formGroupTargetPath.setRequired(true)
        formGroupView.add(formGroupTargetPath)
        val formGroupCommitMessage = FormGroupInputText(
            formGroupView.newChildId(),
            StringResourceModel("gitki.commitmessage"),
            commitMessageModel
        )
        formGroupCommitMessage.addDefaultAjaxInputValidation()
        formGroupCommitMessage.setRequired(true)
        formGroupView.add(formGroupCommitMessage)
    }

    override fun populateFormActions(formActionView: RepeatingView) {
        super.populateFormActions(formActionView)
        formActionView.add(
            formActionView.add(
                SubmitLabelButton(
                    formActionView.newChildId(),
                    StringResourceModel("gitki.move")
                )
            )
        )
        val cancelButton = Label(formActionView.newChildId(), StringResourceModel("gitki.cancel"))
        cancelButton.add(ButtonBehavior())
        cancelButton.add(OnClickScriptBehavior(this.hideScript))
        formActionView.add(cancelButton)
    }

    protected val targetPath: FilePath
        protected get() = targetDirectoryModel.getObject()!!.appendFileName(targetNameModel.getObject()!!)

    protected fun onFileMoved(target: AjaxRequestTarget?, sourcePath: FilePath, targetPath: FilePath) {
        send(this.page, Broadcast.BREADTH, FileMovedEvent(sourcePath, target, targetPath))
    }

    override fun onSubmit(target: AjaxRequestTarget) {
        super.onSubmit(target)
        try {
            gitService!!.moveAndCommit(
                this@MoveFileModal.modelObject!!,
                targetPath,
                getGitkiSession().user!!,
                commitMessageModel.getObject()
            )
        } catch (e: IOException) {
            throw WicketRuntimeException(e)
        } catch (e: GitAPIException) {
            throw WicketRuntimeException(e)
        }
    }

    override fun onAfterSubmit(target: AjaxRequestTarget) {
        super.onAfterSubmit(target)
        onFileMoved(target, this@MoveFileModal.modelObject, targetPath)
    }

    init {
        commitMessageModel = Model.of("Moving " + this.modelObject!!.absoluteString)
        targetNameModel = Model.of(this.modelObject!!.name)
        targetDirectoryModel = Model.of(this.modelObject!!.parent)
    }
}